### ISSUE TITLE
[REGRESSION] `Masonry2` uses `getDerivedStateFromProps` but its initial state is undefined

### ISSUE THREAD

**Issue by wowblvck** [cedricdelpoux/react-responsive-masonry#131]
After the last update (v.2.4.1) i have a regression bug. On the v2.3.0 it's all works good.

![image](https://github.com/user-attachments/assets/9eda3fae-e9ab-48b5-a86e-2af94bf6f217)

**Comment by rchervot** [comment]
Same for me. But I am using Masonry. Also after upgrade to 2.4.1 component started crushing on client.

```
Warning: `Masonry` uses `getDerivedStateFromProps` but its initial state is undefined. This is not recommended. Instead, define the initial state by assigning an object to `this.state` in the constructor of `Masonry`. This ensures that `getDerivedStateFromProps` arguments have a consistent shape.
    at Masonry (webpack-internal:///(ssr)/./node_modules/react-responsive-masonry/es/Masonry/index.js:22:29)
    at div
    at MasonryResponsive (webpack-internal:///(ssr)/./node_modules/react-responsive-masonry/es/ResponsiveMasonry/index.js:49:36)
    at div
```

**Comment by circularchaos** [comment]
Same issue here with 2.4.1. Reverting back to 2.3.0 for now resolved the issue as previously mentioned.

**Comment by wowblvck** [comment]
I want to express my deep gratitude to the author and all those who support this library. Thank you very much for your contribution to its development.

This problem has existed for 2 weeks without the author's attention and since we use this library in our work project, I decided to make my own version of the library based on it and transfer the tools to React 18. I removed all unnecessary tools and tests to cover the features of my project.

This is not advertising or propaganda for switching to my product, this is just a solution for those who also use this library in their projects. I am not going to support or develop it. As soon as I have a little more time, perhaps I will be able to issue a PR. Perhaps the authors of the library will want to apply my solution.

I use Masonry in my projects. We do not use ResponsiveMasonry, so I did not devote time to it and did not test it. Full support for TS.

[react-responsive-masonry (npm)](https://www.npmjs.com/package/@wowblvck/react-responsive-masonry)
[github](https://github.com/wowblvck/react-responsive-masonry)

**Comment by ADTC** [comment]
@wowblvck why was `<Masonry>` changed to `<Index>` in the Readme? Is that an intended change, or a mistake?

Anyway I like that you have converted the project to Typescript and cleaned up a lot. That means I don't need to install `@types/*` package.

(I'm not sure if this particular issue is fixed, but I suppose it is, right?)

**Comment by ADTC** [comment]
@wowblvck BTW for this part:

<img width="740" alt="image" src="https://github.com/user-attachments/assets/33159601-2c30-4910-9eaa-3f8f5d5b73cd">

If you rename the file to `index.tsx` (not `index.ts`) you can simply use `<div>` element tag as it originally already was. It will automagically convert to `React.createElement` so you don't have to make these manual calls.

**Comment by wowblvck** [comment]
> @wowblvck why was `<Masonry>` changed to `<Index>` in the Readme? Is that an intended change, or a mistake?
> 
> Anyway I like that you have converted the project to Typescript and cleaned up a lot. That means I don't need to install `@types/*` package.
> 
> (I'm not sure if this particular issue is fixed, but I suppose it is, right?)

`<Index />` in README.md is a mistake, of course you should use `<Masonry />`


> @wowblvck BTW for this part:
> 
> <img alt="image" width="740" src="https://private-user-images.githubusercontent.com/6047296/387469995-33159601-2c30-4910-9eaa-3f8f5d5b73cd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMjAxMzMsIm5iZiI6MTczMjAxOTgzMywicGF0aCI6Ii82MDQ3Mjk2LzM4NzQ2OTk5NS0zMzE1OTYwMS0yYzMwLTQ5MTAtOWVhYS0zZjhmNWQ1YjczY2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTIzNzEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmZmM2JkYzgzN2RmODZiOTQyZDAxY2I2OWRjNGVkNmRjMDhkNTY3ZmQ0OGQ0ZTg0OTkyNmMxZjg1MTliMjk4MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.NmP7UaYux6WiLtib8W15qH3Bjt_63TogDXoAEtgyf-A">
> If you rename the file to `index.tsx` (not `index.ts`) you can simply use `<div>` element tag as it originally already was. It will automagically convert to `React.createElement` so you don't have to make these manual calls.

Yeah, I know, I just didn't touch that part because I thought it was better that way.

**Comment by ADTC** [comment]
Thanks @wowblvck. BTW could you pinpoint the exact change that specifically fixes the problem of this issue?

> Masonry uses `getDerivedStateFromProps` but its initial state is undefined

Thank you in advance for pointing it out.

I tried looking for it but I'm lost as the commit ["feat: fix masonry bug"](https://github.com/wowblvck/react-responsive-masonry/commit/45ac6cb5cc36ffe6b7d1bdd645d564e0ecd5916e) is showing 40 changed files with 2,694 additions and 25,261 deletions. 😆😆😆

**Comment by ADTC** [comment]
@wowblvck 

> I just didn't touch that part

Hmm, actually I was showing that `<div ... >` changed to `createElement("div", ...)` which is what I felt is an unnecessary change which could be avoided if the files were named correctly to `.tsx` instead of `.ts`.

Anyway, could I ask your help for this? Could you pinpoint the exact change that specifically fixes the problem of this issue? It's hard to track it down as your repository has overhauled the entire project. I want to just apply that fix alone without any other change.

CC @cedricdelpoux

**Comment by cedricdelpoux** [comment]
Hello everyone. Sorry for the delay. I will try to look into this issue soon but PR are welcome if someone want to help. 
@wowblvck if your fork fixed the issue, why not make a PR? Thank you

**Comment by cedricdelpoux** [comment]
Hello, everyone. Sorry for the delay.

I reverted the commits that caused this issue and published an `v2.5.0`.
Tell me if the regression is gone for you.

Thanks

**Comment by ADTC** [comment]
Oh I completely forgot to post this earlier! @cedricdelpoux

I did a bit of testing and found out the fix is actually very simple.  The `class Masonry` needs to have a constructor defined, with this code:

```js
class Masonry extends React.Component {
  constructor() {
    this.state = { columns: [], childRefs: [], hasDistributed: false };
  }

  // rest of the code
}
```

That's all that was needed to be done.

**Comment by cedricdelpoux** [comment]
Ok thanks. I will re add the feature in the next days

**Comment by ADTC** [comment]
> re add the feature

Please check #134 @cedricdelpoux :)

**Comment by cedricdelpoux** [comment]
I just published `2.6.0` with the fix. Thank you for your help 🙂

**Comment by ADTC** [comment]
Happy to say that this bug is truly fixed in `2.6.0` without reverting any code. The error doesn't occur anymore.

Apologies for the panic-induced back and forth reverts I caused @yangchristina 😆
